<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>NEXTFLEX Small RNA · NEXTFLEX Small RNA Anlaysis Documentation</title><script data-outdated-warner src="assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="assets/documenter.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href>NEXTFLEX Small RNA Anlaysis Documentation</a></span></div><form class="docs-search" action="search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li class="is-active"><a class="tocitem" href>NEXTFLEX Small RNA</a><ul class="internal"><li><a class="tocitem" href="#Index"><span>Index</span></a></li><li><a class="tocitem" href="#Functions"><span>Functions</span></a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>NEXTFLEX Small RNA</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>NEXTFLEX Small RNA</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/cookienocreams/small_rna/blob/main/docs/src/index.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="NEXTFLEX-Small-RNA"><a class="docs-heading-anchor" href="#NEXTFLEX-Small-RNA">NEXTFLEX Small RNA</a><a id="NEXTFLEX-Small-RNA-1"></a><a class="docs-heading-anchor-permalink" href="#NEXTFLEX-Small-RNA" title="Permalink"></a></h1><h2 id="Index"><a class="docs-heading-anchor" href="#Index">Index</a><a id="Index-1"></a><a class="docs-heading-anchor-permalink" href="#Index" title="Permalink"></a></h2><ul><li><a href="#small_rna.Config"><code>small_rna.Config</code></a></li><li><a href="#small_rna.MissingReferenceError"><code>small_rna.MissingReferenceError</code></a></li><li><a href="#small_rna.align_with_bowtie2"><code>small_rna.align_with_bowtie2</code></a></li><li><a href="#small_rna.calculate_metrics"><code>small_rna.calculate_metrics</code></a></li><li><a href="#small_rna.calculate_read_length_distribution"><code>small_rna.calculate_read_length_distribution</code></a></li><li><a href="#small_rna.capture_target_files"><code>small_rna.capture_target_files</code></a></li><li><a href="#small_rna.cq_vs_mirna_count_correlation"><code>small_rna.cq_vs_mirna_count_correlation</code></a></li><li><a href="#small_rna.create_target_organism_fasta_file"><code>small_rna.create_target_organism_fasta_file</code></a></li><li><a href="#small_rna.find_common_mirnas"><code>small_rna.find_common_mirnas</code></a></li><li><a href="#small_rna.generate_mirna_counts"><code>small_rna.generate_mirna_counts</code></a></li><li><a href="#small_rna.get_genus_names"><code>small_rna.get_genus_names</code></a></li><li><a href="#small_rna.get_read_q_score!"><code>small_rna.get_read_q_score!</code></a></li><li><a href="#small_rna.mirna_discovery_calculation"><code>small_rna.mirna_discovery_calculation</code></a></li><li><a href="#small_rna.parse_fastqs"><code>small_rna.parse_fastqs</code></a></li><li><a href="#small_rna.plot_clustering"><code>small_rna.plot_clustering</code></a></li><li><a href="#small_rna.plot_cq_linear_regression"><code>small_rna.plot_cq_linear_regression</code></a></li><li><a href="#small_rna.plot_fragment_lengths"><code>small_rna.plot_fragment_lengths</code></a></li><li><a href="#small_rna.plot_metrics"><code>small_rna.plot_metrics</code></a></li><li><a href="#small_rna.plot_mirna_counts"><code>small_rna.plot_mirna_counts</code></a></li><li><a href="#small_rna.progress_bar_update"><code>small_rna.progress_bar_update</code></a></li><li><a href="#small_rna.remove_files"><code>small_rna.remove_files</code></a></li><li><a href="#small_rna.remove_intermediate_files"><code>small_rna.remove_intermediate_files</code></a></li><li><a href="#small_rna.trim_adapters"><code>small_rna.trim_adapters</code></a></li><li><a href="#small_rna.write_common_mirna_file"><code>small_rna.write_common_mirna_file</code></a></li></ul><h2 id="Functions"><a class="docs-heading-anchor" href="#Functions">Functions</a><a id="Functions-1"></a><a class="docs-heading-anchor-permalink" href="#Functions" title="Permalink"></a></h2><p>Documentation for small_rna.jl</p><article class="docstring"><header><a class="docstring-binding" id="small_rna.MissingReferenceError" href="#small_rna.MissingReferenceError"><code>small_rna.MissingReferenceError</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">struct MissingReferenceError
    message::String
end</code></pre><p>A structure that represents a custom exception for missing reference.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/cookienocreams/small_rna/blob/7fc206830797cb2c9e6ebe7ba62184da4c209375/src/small_rna.jl#L53-L59">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="small_rna.Config" href="#small_rna.Config"><code>small_rna.Config</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">struct Config
    library::Union{String, Nothing}
    fasta::Union{String, Nothing}
    organism::Union{String, Nothing}
    threads::Int
end</code></pre><p>A structure that stores the command-line arguments.</p><p><strong>Functionality</strong></p><ul><li>Contains a check to ensure that a miRNA fasta file is passed as an argument. It is was</li></ul><p>not, an error will be returned. This is to reduce the chances of an alignment errors downstream.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/cookienocreams/small_rna/blob/7fc206830797cb2c9e6ebe7ba62184da4c209375/src/small_rna.jl#L64-L78">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="small_rna.get_genus_names" href="#small_rna.get_genus_names"><code>small_rna.get_genus_names</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Create a file with a list of unique species names given a miRNA mature or hairpin fasta.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/cookienocreams/small_rna/blob/7fc206830797cb2c9e6ebe7ba62184da4c209375/src/small_rna.jl#L99-L101">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="small_rna.create_target_organism_fasta_file" href="#small_rna.create_target_organism_fasta_file"><code>small_rna.create_target_organism_fasta_file</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Function takes in mirBase mature miRNA fasta file with data from all available organisms and pulls out only the miRNA data pertaining to the target organism.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/cookienocreams/small_rna/blob/7fc206830797cb2c9e6ebe7ba62184da4c209375/src/small_rna.jl#L117-L120">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="small_rna.capture_target_files" href="#small_rna.capture_target_files"><code>small_rna.capture_target_files</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">capture_target_files(files_to_capture::AbstractString, directory::AbstractString=&quot;.&quot;)</code></pre><p>List all files in a directory. </p><p>Check to see if each file contains the target file&#39;s string. </p><p><strong>Example</strong></p><pre><code class="language-julia hljs">julia&gt; capture_target_files(&quot;.txt&quot;)
3-element Vector{String}:
 &quot;file1.txt&quot;
 &quot;file2.txt&quot;
 &quot;file3.txt&quot;</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/cookienocreams/small_rna/blob/7fc206830797cb2c9e6ebe7ba62184da4c209375/src/small_rna.jl#L181-L196">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="small_rna.progress_bar_update" href="#small_rna.progress_bar_update"><code>small_rna.progress_bar_update</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Update progress bar on the command line.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/cookienocreams/small_rna/blob/7fc206830797cb2c9e6ebe7ba62184da4c209375/src/small_rna.jl#L201-L203">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="small_rna.get_read_q_score!" href="#small_rna.get_read_q_score!"><code>small_rna.get_read_q_score!</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">get_read_q_score!(line::String, q_score_list::Vector{Number})</code></pre><p>Calculate the quality score for a given quality string.</p><p>This function accepts ASCII-encoded quality score and produces the average q score  using a scalar value that estimates the &#39;total error probability&#39; for that record. It means that if you want to calculate average quality score, you don&#39;t just sum up all the phred scores and find the average, but rather consider the  probability distribution of the scores.</p><p>The Phred score for a base <code>Q</code> is calculated as <code>Q = ASCII value of quality score - 33</code>. The error probability <code>P</code> for that base is then calculated as <code>P = 10^(-Q/10)</code>. Then these probabilities are summed up for all the bases to derive total error.</p><p><strong>Arguments</strong></p><p><code>q_score</code> - The quality scores of the current line. <code>q_score_list</code> - The vector containing all calculated quality scores.</p><p><strong>Returns</strong></p><p>The updated q score list the line&#39;s average quality score.</p><p><strong>Reference</strong></p><p>Illumina&#39;s explanation on quality scores and their corresponding error rates: <a href="https://www.illumina.com/science/technology/next-generation-sequencing/plan-experiments/quality-scores.html">https://www.illumina.com/science/technology/next-generation-sequencing/plan-experiments/quality-scores.html</a></p><p><strong>Example</strong></p><pre><code class="language-julia hljs">julia&gt; get_read_q_score!(&quot;FGFEEE&lt;FC&lt;FGGGGGGGGGGGFFGFG8&lt;@8CFFGF8EFGGCGFGGGGGGFG&quot;, [36.2, 35.9])
3-element Vector{Float64}:
 36.2
 35.9
 35.7</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/cookienocreams/small_rna/blob/7fc206830797cb2c9e6ebe7ba62184da4c209375/src/small_rna.jl#L218-L255">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="small_rna.parse_fastqs" href="#small_rna.parse_fastqs"><code>small_rna.parse_fastqs</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">parse_fastqs(fastqs::Vector{String}, sample_names::Vector{SubString{String}}, library::String)</code></pre><p>Determine the library type, calculate read counts, percent dimers, and average Q scores  for a given set of fastq files.</p><p><strong>Parameters</strong></p><ul><li><code>fastqs</code>: A vector of filenames for the fastq files.</li><li><code>sample_names</code>: A vector of sample names corresponding to the fastq files.</li><li><code>library</code>: The type of library (&quot;v3&quot; or &quot;v4&quot;).</li></ul><p><strong>Returns</strong></p><p>Three dictionaries containing read counts, percent dimers, and average Q scores  indexed by sample name.</p><p><strong>Example</strong></p><pre><code class="language-julia hljs">julia&gt; parse_fastqs([&quot;sample1.fastq.gz&quot;, &quot;sample2.fastq.gz&quot;, &quot;sample3.fastq.gz&quot;], 
                    [&quot;sample1&quot;, &quot;sample2&quot;, &quot;sample3&quot;], &quot;v4&quot;)
# Returns:
# - A dictionary of read counts.
# - A dictionary of percent dimers.
# - A dictionary of average Q scores.</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/cookienocreams/small_rna/blob/7fc206830797cb2c9e6ebe7ba62184da4c209375/src/small_rna.jl#L275-L299">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="small_rna.trim_adapters" href="#small_rna.trim_adapters"><code>small_rna.trim_adapters</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">trim_adapters(fastqs::Vector{String}, library::String, sample_names::Vector{SubString{String}})</code></pre><p>Trim the 3&#39; adapter from each read in the provided fastq files based on the given library type.</p><p><strong>Library Adapters:</strong></p><ul><li><p>V3 Read 1:  5&#39; Adapter: GATCGTCGGACTGTAGAACTCTGAACNNNN   3&#39; Adapter: NNNNTGGAATTCTCGGGTGCCAAGG</p></li><li><p>V4 Read 1:  5&#39; Adapter: AGATCGGAAGAGCGTCGTGTAGGGAAAGA  3&#39; Adapter: TGGAATTCTCGGGTGCCAAGG </p></li></ul><p>3&#39; bases with a quality score below 20 are trimmed. Reads shorter than 16 bases or  those that weren&#39;t trimmed are discarded.</p><p><strong>Parameters</strong></p><ul><li><code>fastqs</code>: A vector of filenames for the fastq files.</li><li><code>library</code>: The type of library (&quot;v3&quot; or &quot;v4&quot;).</li><li><code>sample_names</code>: A vector of sample names corresponding to the fastq files.</li></ul><p><strong>Returns</strong></p><p>A list of filenames for the trimmed fastq files.</p><p><strong>Example</strong></p><pre><code class="language-julia hljs">julia&gt; trim_adapters([&quot;sub_sample1.fastq&quot;, &quot;sub_sample2.fastq&quot;, &quot;sub_sample3.fastq&quot;], 
                    &quot;v4&quot;, 
                    [&quot;sample1&quot;, &quot;sample2&quot;, &quot;sample3&quot;])
# Returns:
[&quot;sample1.cut.fastq&quot;, &quot;sample2.cut.fastq&quot;, &quot;sample3.cut.fastq&quot;]</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/cookienocreams/small_rna/blob/7fc206830797cb2c9e6ebe7ba62184da4c209375/src/small_rna.jl#L353-L386">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="small_rna.generate_mirna_counts" href="#small_rna.generate_mirna_counts"><code>small_rna.generate_mirna_counts</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">generate_mirna_counts(input_sam_file::AbstractString)</code></pre><p>This function creates a DataFrame containing all aligned miRNA and the number of times  they appear.</p><p><strong>Returns</strong></p><p>A DataFrame with the following columns:</p><ul><li><code>name</code>: The name of the miRNA.</li><li><code>count</code>: The number of times the miRNA appears in the SAM file.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/cookienocreams/small_rna/blob/7fc206830797cb2c9e6ebe7ba62184da4c209375/src/small_rna.jl#L422-L434">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="small_rna.mirna_discovery_calculation" href="#small_rna.mirna_discovery_calculation"><code>small_rna.mirna_discovery_calculation</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">mirna_discovery_calculation(trimmed_fastqs::Vector{String}, sample_names::Vector{SubString{String}}, bowtie2_reference_name::String)</code></pre><p>Align trimmed fastq files to a given human miRNA bowtie2 reference and calculate the number of unique miRNAs present in each sample.</p><p>The function returns:</p><ol><li>A <code>Vector</code> of DataFrames where each DataFrame contains the counts of miRNAs for a specific sample.</li><li>A <code>Vector</code> of filenames of the SAM files generated during the alignment.</li></ol><p><strong>Parameters</strong></p><ul><li><code>trimmed_fastqs</code>: A vector of filenames for the trimmed fastq files.</li><li><code>sample_names</code>: A vector of sample names corresponding to the fastq files.</li><li><code>bowtie2_reference_name</code>: The name of the bowtie2 reference against which the alignment should be done.</li><li><code>cores</code>: The number of threads to use for alignment.</li></ul><p><strong>Returns</strong></p><ol><li>A <code>Vector</code> of DataFrames. Each DataFrame has two columns: <code>name</code> representing the miRNA name and <code>count</code> representing the number of times the miRNA was observed in the sample.</li><li>A <code>Vector</code> of filenames for the SAM files generated during the alignment.</li></ol><p><strong>Example</strong></p><pre><code class="language-julia hljs">julia&gt; mirna_discovery_calculation([&quot;sample1.cut.fastq&quot;, &quot;sample2.cut.fastq&quot;, &quot;sample3.cut.fastq&quot;], 
                                   [&quot;sample1&quot;, &quot;sample2&quot;, &quot;sample3&quot;], 
                                   &quot;human_miRNA_reference&quot;)
# Returns:
# ([469x2 DataFrame
#    Row │ name             count
#    │ String           Int64
#    ─────┼────────────────────────
#      1 │ hsa-miR-1307-3p    8548
#      2 │ hsa-miR-425-5p     3221
#      3 │ hsa-miR-488-5p     2951
#   ...], 
#  [&quot;sample1.miRNA.sam&quot;, &quot;sample2.miRNA.sam&quot;, &quot;sample3.miRNA.sam&quot;])</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/cookienocreams/small_rna/blob/7fc206830797cb2c9e6ebe7ba62184da4c209375/src/small_rna.jl#L471-L506">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="small_rna.plot_mirna_counts" href="#small_rna.plot_mirna_counts"><code>small_rna.plot_mirna_counts</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">plot_mirna_counts(mirna_counts_dfs::Vector{DataFrame}, sample_names::Vector{SubString{String}})</code></pre><p>Plot counts of the number of unique miRNA each sample aligned to. The function returns the names of the .csv files generated.</p><p><strong>Example</strong></p><pre><code class="language-julia hljs">julia&gt; plot_mirna_counts([&quot;sample1.hairpin.hist&quot;,&quot;sample2.hairpin.hist&quot;,&quot;sample3.hairpin.hist&quot;], 
                         [&quot;sample1&quot;, &quot;sample2&quot;, &quot;sample3&quot;])
[&quot;sample1_miRNA_counts.csv&quot;, &quot;sample2_miRNA_counts.csv&quot;, &quot;sample3_miRNA_counts.csv&quot;]</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/cookienocreams/small_rna/blob/7fc206830797cb2c9e6ebe7ba62184da4c209375/src/small_rna.jl#L553-L564">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="small_rna.calculate_read_length_distribution" href="#small_rna.calculate_read_length_distribution"><code>small_rna.calculate_read_length_distribution</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Use aligned SAM files to create a read length distribution file.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/cookienocreams/small_rna/blob/7fc206830797cb2c9e6ebe7ba62184da4c209375/src/small_rna.jl#L633-L635">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="small_rna.plot_fragment_lengths" href="#small_rna.plot_fragment_lengths"><code>small_rna.plot_fragment_lengths</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Create barplot of each sample&#39;s fragment lengths based on the number of reads found at each length</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/cookienocreams/small_rna/blob/7fc206830797cb2c9e6ebe7ba62184da4c209375/src/small_rna.jl#L687-L689">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="small_rna.align_with_bowtie2" href="#small_rna.align_with_bowtie2"><code>small_rna.align_with_bowtie2</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">align_with_bowtie2(fastq_file::String, read_count_dict::Dict{String, Int64})</code></pre><p>Align fastq with specified reference RNA types.</p><p>Divide reads aligned with total reads to calculate percent alignment. Store  alignment information in a dictionary.</p><p><strong>Example</strong></p><pre><code class="language-julia hljs">julia&gt; align_with_bowtie2(&quot;sample1.cut.fastq&quot;
                        , Dict(&quot;sample1&quot; =&gt; 6390314, &quot;sample2&quot; =&gt; 5000000, &quot;sample3&quot; =&gt; 7052928))
Dict{String, Float64} with 5 entries:
  &quot;miRNA&quot; =&gt; 65.0
  &quot;tRNA&quot; =&gt; 4.2
  &quot;piRNA&quot; =&gt; 1.2
  &quot;snoRNA&quot; =&gt; 1.9
  &quot;rRNA&quot; =&gt; 3.3</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/cookienocreams/small_rna/blob/7fc206830797cb2c9e6ebe7ba62184da4c209375/src/small_rna.jl#L739-L758">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="small_rna.calculate_metrics" href="#small_rna.calculate_metrics"><code>small_rna.calculate_metrics</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Calculate the bowtie2 alignment metrics for all samples. Return filename for plotting.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/cookienocreams/small_rna/blob/7fc206830797cb2c9e6ebe7ba62184da4c209375/src/small_rna.jl#L840-L842">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="small_rna.plot_metrics" href="#small_rna.plot_metrics"><code>small_rna.plot_metrics</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Create barplot of each sample&#39;s RNA aligment metrics.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/cookienocreams/small_rna/blob/7fc206830797cb2c9e6ebe7ba62184da4c209375/src/small_rna.jl#L971-L973">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="small_rna.find_common_mirnas" href="#small_rna.find_common_mirnas"><code>small_rna.find_common_mirnas</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">find_common_mirnas(mirna_counts_files::Vector{String}
                        , read_count_dict::Dict{String, Int64}
                        , sample_names::Vector{SubString{String}}
                        )</code></pre><p>Return the miRNAs present in all samples, together with their counts and reads per million reads (RPM).</p><p><strong>Arguments</strong></p><ul><li><code>mirna_counts_files</code>: A vector of files containing sample miRNA counts.</li><li><code>read_count_dict</code>: A dictionary where keys are sample names and values are the total number of reads </li></ul><p>for the corresponding sample.</p><ul><li><code>sample_names</code>: A vector of sample names.</li></ul><p><strong>Returns</strong></p><ul><li>A Tuple of three items:<ol><li>A list of miRNAs that are common to all samples.</li><li>A list of dictionaries where each dictionary contains the miRNA counts for one sample.</li><li>A list of dictionaries where each dictionary contains the RPM values for the miRNAs in one sample.</li></ol></li></ul><p><strong>Example</strong></p><pre><code class="language-julia hljs">julia&gt; find_common_mirnas([&quot;sample1_miRNA_counts.csv&quot;, &quot;sample2_miRNA_counts.csv&quot;]...,
                        , Dict(&quot;sample1&quot; =&gt; 6390314, &quot;sample2&quot; =&gt; 5000000, &quot;sample3&quot; =&gt; 7052928)
                        , [&quot;sample1&quot;, &quot;sample2&quot;, &quot;sample3&quot;])</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/cookienocreams/small_rna/blob/7fc206830797cb2c9e6ebe7ba62184da4c209375/src/small_rna.jl#L1036-L1062">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="small_rna.write_common_mirna_file" href="#small_rna.write_common_mirna_file"><code>small_rna.write_common_mirna_file</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Write miRNAs all samples have in common to output file.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/cookienocreams/small_rna/blob/7fc206830797cb2c9e6ebe7ba62184da4c209375/src/small_rna.jl#L1109-L1111">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="small_rna.plot_clustering" href="#small_rna.plot_clustering"><code>small_rna.plot_clustering</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">plot_clustering(Common_miRNA_File::String)</code></pre><p>Generates Principal Component Analysis (PCA) and Uniform Manifold Approximation and  Projection (UMAP) plots for libraries given a common miRNA RPM file.</p><p>This function performs the following steps:</p><ol><li>Reads the common miRNA RPM file and creates a DataFrame.</li><li>Extracts sample names and miRNA names.</li><li>Ensures that there are at least two samples and more than one miRNA present in the dataset.</li><li>Transforms the data using PCA and prepares it for plotting.</li><li>If there are at least two principal components, plots the PCA and saves it as &quot;common<em>miRNA</em>PCA.png&quot;.</li><li>If there are at least three principal components, performs UMAP dimensionality reduction and clustering using K-medoids.</li><li>Plots the UMAP and saves it as &quot;common<em>miRNA</em>UMAP.png&quot;.</li><li>Writes the PCA and UMAP information to separate CSV files for cluster tracking.</li></ol><p><strong>Arguments</strong></p><ul><li><code>Common_miRNA_File::String</code>: Path to the common miRNA RPM file.</li></ul><p><strong>Outputs</strong></p><ul><li>Create and saves PCA and UMAP plots.</li><li>Write PCA and UMAP information.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/cookienocreams/small_rna/blob/7fc206830797cb2c9e6ebe7ba62184da4c209375/src/small_rna.jl#L1156-L1178">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="small_rna.cq_vs_mirna_count_correlation" href="#small_rna.cq_vs_mirna_count_correlation"><code>small_rna.cq_vs_mirna_count_correlation</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">cq_vs_mirna_count_correlation(qPCR_Data_File::String,
                              Common_miRNA_RPM_File::String,
                              Common_miRNA_File::String,
                              sample_names::Vector{SubString{String}}) -&gt; DataFrame</code></pre><p>Calculate the Pearson&#39;s correlation coefficient between qPCR data and common miRNA counts.</p><p>Functionality:</p><ul><li>Filters the qPCR data to exclude measurements without a quantification cycle (Cq) value.</li><li>Intersects the filtered qPCR data with the miRNAs from the common miRNA files.</li><li>Computes the average Cq value for each miRNA across replicates.</li><li>Determines the Pearson&#39;s correlation coefficient between the qPCR data and the common miRNA counts.</li><li>If the Pearson&#39;s coefficient is &lt; -0.6 for any sample, a linear regression is plotted for that sample.</li></ul><p><strong>Parameters:</strong></p><ul><li><code>qPCR_Data_File</code>: Path to the file containing qPCR data.</li><li><code>Common_miRNA_RPM_File</code>: Path to the file containing RPM data for common miRNAs.</li><li><code>Common_miRNA_File</code>: Path to the file listing common miRNAs.</li><li><code>sample_names</code>: List of sample names to process.</li></ul><p><strong>Returns:</strong></p><ul><li>A DataFrame containing miRNA names, their average Cq values, and RPM counts for each sample.</li></ul><p><strong>Example:</strong></p><pre><code class="language-julia hljs">julia&gt; cq_vs_mirna_count_correlation(&quot;qpcr_raw_data.csv&quot;, &quot;Common_RPM_miRNAs.csv&quot;,
                                     &quot;Common_miRNAs.csv&quot;, [&quot;sample1&quot;, &quot;sample2&quot;, &quot;sample3&quot;])                   
# Outputs:
# A DataFrame with columns for miRNA, Cq, and RPM values for each sample.
# A CSV file named &quot;Sample_correlation_coefficients.csv&quot; containing sample names and their correlation coefficients.</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/cookienocreams/small_rna/blob/7fc206830797cb2c9e6ebe7ba62184da4c209375/src/small_rna.jl#L1254-L1286">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="small_rna.plot_cq_linear_regression" href="#small_rna.plot_cq_linear_regression"><code>small_rna.plot_cq_linear_regression</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Make linear regression plot for samples correlated with the qPCR data.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/cookienocreams/small_rna/blob/7fc206830797cb2c9e6ebe7ba62184da4c209375/src/small_rna.jl#L1349-L1351">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="small_rna.remove_files" href="#small_rna.remove_files"><code>small_rna.remove_files</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Spot remove unecessary intermediate files.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/cookienocreams/small_rna/blob/7fc206830797cb2c9e6ebe7ba62184da4c209375/src/small_rna.jl#L1405-L1407">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="small_rna.remove_intermediate_files" href="#small_rna.remove_intermediate_files"><code>small_rna.remove_intermediate_files</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Remove all intermediate files.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/cookienocreams/small_rna/blob/7fc206830797cb2c9e6ebe7ba62184da4c209375/src/small_rna.jl#L1418-L1420">source</a></section></article></article><nav class="docs-footer"><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.25 on <span class="colophon-date" title="Sunday 27 August 2023 16:22">Sunday 27 August 2023</span>. Using Julia version 1.9.3.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
